doozerlib/runtime.py

```diff
@@ -644,56 +644,44 @@ def resolve_source(self, parent, meta):
 
         url = source_details["url"]
         branches = source_details['branch']
-        self.logger.info("Cloning source '%s' from %s as specified by group into: %s" % (alias, url, source_dir))
-        exectools.cmd_assert(
-            cmd=["git", "clone", url, source_dir],
-            retries=3,
-            on_retry=["rm", "-rf", source_dir],
-        )
+
         stage_branch = branches.get('stage', None)
         fallback_branch = branches.get("fallback", None)
         found = False
-        with Dir(source_dir):
-            if self.stage and stage_branch:
-                self.logger.info('Normal branch overridden by --stage option, using "{}"'.format(stage_branch))
-                branch = stage_branch
-            else:
-                branch = branches["target"]
-            self.logger.info("Attempting to checkout source '%s' branch %s in: %s" % (alias, branch, source_dir))
 
-            if branch != "master":
-                rc, out, err = exectools.cmd_gather(["git", "checkout", "-b", branch, "origin/%s" % branch])
-            else:
-                rc = 0
+        if self.stage and stage_branch:
+            self.logger.info('Normal branch overridden by --stage option, using "{}"'.format(stage_branch))
+            branch = stage_branch
+        else:
+            branch = branches["target"]
 
-            if rc == 0:
-                found = True
-            else:
+        base_clone = "git clone -b {} --single-branch {} --depth 1 {}"
+        clone_target = base_clone.format(branch, url, source_dir)
+        clone_fallback = base_clone.format(fallback_branch, url, source_dir)
+        retries = 3
+        while retries:
+            try:
+                self.logger.info("Attempting to checkout source '%s' branch %s in: %s" % (alias, branch, source_dir))
+                exectools.cmd_assert(clone_target)
+            except IOError:
                 if self.stage and stage_branch:
                     raise DoozerFatalError('--stage option specified and no stage branch named "{}" exists for {}|{}'.format(stage_branch, alias, url))
-                elif fallback_branch is not None:
+                try:
```

Comment:
Also it bothers me a bit to go right from an attempted clone to the fallback, and without cleaning up. We don't know why the first one failed... maybe it was part-way cloned and then there was a network bobble. We would probably want to rm the directory and try again. We would probably not want to try the backup without checking to see if the initial branch even existed. Maybe a git remote-ls first to decide what can be cloned, then you just have a single thing to clone (and retry)?